src/waveform/renderers/qopengl/waveformrendererfiltered.cpp

```diff
@@ -0,0 +1,255 @@
+#include "waveform/renderers/qopengl/waveformrendererfiltered.h"
+
+#include <iostream>
+
+#include "track/track.h"
+#include "util/math.h"
+#include "waveform/waveform.h"
+#include "waveform/waveformwidgetfactory.h"
+#include "waveform/widgets/qopengl/waveformwidget.h"
+#include "widget/wskincolor.h"
+#include "widget/wwidget.h"
+
+using namespace qopengl;
+
+WaveformRendererFiltered::WaveformRendererFiltered(
+        WaveformWidgetRenderer* waveformWidget)
+        : WaveformRendererSignalBase(waveformWidget) {
+}
+
+WaveformRendererFiltered::~WaveformRendererFiltered() {
+}
+
+void WaveformRendererFiltered::onSetup(const QDomNode& node) {
+    Q_UNUSED(node);
+}
+
+void WaveformRendererFiltered::initializeGL() {
+    QString vertexShaderCode = QStringLiteral(R"--(
+uniform mat4 matrix;
+attribute vec4 position;
+void main()
+{
+    gl_Position = matrix * position;
+}
+)--");
+
+    QString fragmentShaderCode = QStringLiteral(R"--(
+uniform vec4 color;
+void main()
+{
+    gl_FragColor = color;
+}
+)--");
+
+    initShaders(vertexShaderCode, fragmentShaderCode);
+}
+
+inline void WaveformRendererFiltered::addRectangle(
+        QVector<float>& lines,
```

Comment:
mixxx follows the convention to not use mutable references for outparameters. use `gsl::not_null<QVector<float>*>` instead.

Reply:
I changed the API slightly to index a member var instead.